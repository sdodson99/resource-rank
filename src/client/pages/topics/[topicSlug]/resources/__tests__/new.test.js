import { render, screen, waitFor } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import getTopicBySlug from '@/services/topics/graphql-topic-by-slug-service';
import { when } from 'jest-when';
import NewTopicResource, { Page, getServerSideProps } from '../new.page';
import useResourceCreator from '@/hooks/resources/use-resource-creator';
import useTopicResourceCreator from '@/hooks/topics/use-topic-resource-creator';
import useNavigate from '@/hooks/use-navigate';
import ResourceExistsError from '@/errors/resource-exists-error';
import withApp from '@/test-utils/with-app';
import renderer from 'react-test-renderer';
import userEvent from '@testing-library/user-event';

jest.mock('@/hooks/use-navigate');
jest.mock('@/hooks/resources/use-resource-creator');
jest.mock('@/hooks/topics/use-topic-resource-creator');
jest.mock('@/services/topics/graphql-topic-by-slug-service');

describe('<NewTopicResource />', () => {
  describe('page', () => {
    let topicId;
    let topicName;
    let topicSlug;
    let props;

    beforeEach(() => {
      useResourceCreator.mockReturnValue({});
      useTopicResourceCreator.mockReturnValue({});

      topicId = '123';
      topicName = 'topic-name';
      topicSlug = 'topic-slug';

      props = {
        topicId,
        topicName,
        topicSlug,
      };
    });

    afterEach(() => {
      useNavigate.mockReset();
      useTopicResourceCreator.mockReset();
      useResourceCreator.mockReset();
    });

    it('should mount', () => {
      render(withApp(Page, props));

      const page = screen.getByTestId('NewTopicResourcePage');

      expect(page).toBeInTheDocument();
    });

    describe('on valid submit', () => {
      function submitForm() {
        const submitButton = screen.getByText('Submit');
        submitButton.click();
      }

      function renderAndSubmit() {
        render(withApp(Page, props));

        userEvent.type(screen.getByLabelText('Name'), 'name');
        userEvent.type(screen.getByLabelText('Link'), 'test.com');

        submitForm();
      }

      let mockCreateResource;
      let mockCreateTopicResource;

      let resourceId;
      let resourceSlug;

      beforeEach(() => {
        mockCreateResource = jest.fn();
        useResourceCreator.mockReturnValue({
          createResource: mockCreateResource,
        });

        mockCreateTopicResource = jest.fn();
        useTopicResourceCreator.mockReturnValue({
          createTopicResource: mockCreateTopicResource,
        });

        resourceId = '789';
        resourceSlug = 'resource-slug';
      });

      it('should renavigate to created topic resource if successful', async () => {
        const mockNavigate = jest.fn();
        useNavigate.mockReturnValue(mockNavigate);
        mockCreateResource.mockReturnValue({
          id: resourceId,
          slug: resourceSlug,
        });
        mockCreateTopicResource.mockReturnValue(true);

        renderAndSubmit();

        await waitFor(() => {
          expect(mockNavigate).toBeCalledWith({
            pathname: `/topics/topic-slug/resources/resource-slug`,
            query: {
              new: true,
            },
          });
        });
      });

      it('should display create topic resource error if failed to create topic resource', async () => {
        mockCreateResource.mockReturnValue({
          id: resourceId,
          slug: resourceSlug,
        });
        mockCreateTopicResource.mockReturnValue(false);

        renderAndSubmit();

        await waitFor(() => {
          const errorMessage = screen.getByText(
            'Failed to create topic resource.'
          );
          expect(errorMessage).toBeInTheDocument();
        });
      });

      it('should set name exists error if resource name already exists', async () => {
        mockCreateResource.mockImplementation(() => {
          throw new ResourceExistsError();
        });
        renderAndSubmit();

        await waitFor(() => {
          const errorMessage = screen.getByText('Name already exists.');
          expect(errorMessage).toBeInTheDocument();
        });
      });

      it('should set create resource error if failed to create resource', async () => {
        mockCreateResource.mockImplementation(() => {
          throw new Error();
        });

        renderAndSubmit();

        await waitFor(() => {
          const errorMessage = screen.getByText('Failed to create resource.');
          expect(errorMessage).toBeInTheDocument();
        });
      });

      it('should clear generic error on invalid submit', async () => {
        mockCreateResource.mockReturnValue({
          id: resourceId,
          slug: resourceSlug,
        });
        mockCreateTopicResource.mockReturnValue(false);
        renderAndSubmit();
        await waitFor(() => {
          const errorMessage = screen.getByText(
            'Failed to create topic resource.'
          );
          expect(errorMessage).toBeInTheDocument();
        });
        mockCreateTopicResource.mockReturnValue(true);

        submitForm();

        await waitFor(() => {
          const errorMessage = screen.queryByText(
            'Failed to create topic resource.'
          );
          expect(errorMessage).toBeNull();
        });
      });
    });

    it('should render correctly', () => {
      const page = renderer.create(withApp(Page, props)).toJSON();

      expect(page).toMatchSnapshot();
    });
  });

  describe('HOC page', () => {
    it('should require authentication', () => {
      const page = renderer
        .create(withApp(NewTopicResource, {}, { mock: 'unauthenticated' }))
        .toJSON();

      expect(page).toMatchSnapshot();
    });
  });

  describe('getServerSideProps', () => {
    let req;
    let params;
    let query;
    let mock;
    let topicSlug;

    beforeEach(() => {
      mock = 'mock';
      query = {
        mock,
      };
      req = {};
      topicSlug = 'topic-name';
      params = {
        topicSlug,
      };
    });

    afterEach(() => {
      getTopicBySlug.mockReset();
    });

    it('should return not found if topic query fails', async () => {
      when(getTopicBySlug)
        .calledWith(topicSlug, { mock })
        .mockImplementation(() => {
          throw new Error();
        });

      const { notFound } = await getServerSideProps({ req, params, query });

      expect(notFound).toBeTruthy();
    });

    describe('when successful', () => {
      let id;
      let name;
      let slug;

      beforeEach(() => {
        id = '123';
        name = 'name';
        slug = 'slug';
      });

      it('should return topic props', async () => {
        when(getTopicBySlug).calledWith(topicSlug, { mock }).mockReturnValue({
          id,
          name,
          slug,
        });

        const { props } = await getServerSideProps({ req, params, query });

        expect(props).toEqual({
          topicId: id,
          topicName: name,
          topicSlug: slug,
        });
      });
    });
  });
});
