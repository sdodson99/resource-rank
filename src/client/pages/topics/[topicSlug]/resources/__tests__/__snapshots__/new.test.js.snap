// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<NewTopicResource /> HOC page should require authentication 1`] = `
<div
  className="Layout"
  data-testid="Layout"
>
  <div
    className="Page"
  >
    <header
      className="Header"
      data-testid="Header"
    >
      <div
        className="content-container"
      >
        <div
          className="sm:flex items-center text-center sm:text-left"
        >
          <div
            className="flex-grow flex justify-center sm:justify-start"
          >
            <a
              href="/?mock=unauthenticated"
              onClick={[Function]}
              onMouseEnter={[Function]}
            >
              <img
                alt="Resource Rank Logo"
                height="50"
                src="/img/logo.svg"
                width="50"
              />
            </a>
          </div>
          <div
            className="NavItem NavLink ActiveNavLink"
            data-testid="ActiveLink"
          >
            Home
          </div>
          <div
            className="NavItem"
            data-testid="TopicsNavItem"
          >
            <div
              className="NavLink"
              data-testid="ActiveLink"
            >
              <a
                className=""
                href="/topics?mock=unauthenticated"
                onClick={[Function]}
                onMouseEnter={[Function]}
              >
                Topics
              </a>
            </div>
          </div>
          <div
            className="NavItem"
          >
            <button
              className="btn btn-primary"
              data-testid="LoadingButton"
              onClick={[Function]}
            >
              Login
            </button>
          </div>
        </div>
      </div>
    </header>
    <div>
      <div
        className="mt-8 text-center text-lg"
      >
        You must login to view this page.
      </div>
    </div>
  </div>
  <footer
    className="Footer"
    data-testid="Footer"
  >
    <div
      className="content-container"
    >
      <div
        className="sm:flex text-center sm:text-left"
      >
        <div
          className="sm:order-2 sm:ml-10"
        >
          <h1
            className="text-4xl"
          >
            Our Mission
          </h1>
          <p
            className="py-3"
          >
            is to 
            <strong>
              simplify
            </strong>
             your learning journey. Learning can be 
            <strong>
              difficult
            </strong>
            , especially when starting. There are
             
            <strong>
              millions
            </strong>
             of educational courses, videos, and books
             
            <strong>
              scattered
            </strong>
             across the Internet.
          </p>
          <p
            className="py-3"
          >
            Resource Rankâ€™s goal is to 
            <strong>
              consolidate
            </strong>
             all of these resources so that you can 
            <strong>
              quickly
            </strong>
             find the
             
            <strong>
              best
            </strong>
             resources for what you are learning.
          </p>
        </div>
        <div
          className="sm:order-1 mt-10 sm:mt-0 flex items-start justify-center sm:justify-start"
        >
          <img
            alt="Resource Rank Logo"
            className="max-w-none"
            height="75"
            src="/img/logo.svg"
            width="75"
          />
        </div>
      </div>
    </div>
  </footer>
</div>
`;

exports[`<NewTopicResource /> page should render correctly 1`] = `
<div
  data-testid="NewTopicResourcePage"
>
  <div
    className="BreadcrumbLayout"
    data-testid="BreadcrumbLayout"
  >
    <div
      className="Layout"
      data-testid="Layout"
    >
      <div
        className="Page"
      >
        <header
          className="Header"
          data-testid="Header"
        >
          <div
            className="content-container"
          >
            <div
              className="sm:flex items-center text-center sm:text-left"
            >
              <div
                className="flex-grow flex justify-center sm:justify-start"
              >
                <a
                  href="/?mock=standard"
                  onClick={[Function]}
                  onMouseEnter={[Function]}
                >
                  <img
                    alt="Resource Rank Logo"
                    height="50"
                    src="/img/logo.svg"
                    width="50"
                  />
                </a>
              </div>
              <div
                className="NavItem NavLink ActiveNavLink"
                data-testid="ActiveLink"
              >
                Home
              </div>
              <div
                className="NavItem"
                data-testid="TopicsNavItem"
              >
                <div
                  className="NavLink"
                  data-testid="ActiveLink"
                >
                  <a
                    className=""
                    href="/topics?mock=standard"
                    onClick={[Function]}
                    onMouseEnter={[Function]}
                  >
                    Topics
                  </a>
                </div>
              </div>
              <div
                className="NavItem"
              >
                <button
                  className="btn btn-primary"
                  data-testid="LoadingButton"
                  onClick={[Function]}
                >
                  Login
                </button>
              </div>
            </div>
          </div>
        </header>
        <div>
          <div
            className="content-container pb-10"
          >
            <nav
              aria-label="breadcrumb"
              className="BreadcrumbContent"
            >
              <ol
                className="flex flex-wrap"
              >
                <li
                  className="flex"
                >
                  <div
                    className="BreadcrumbItem"
                    data-testid="ActiveLink"
                  >
                    <a
                      className="hyperlink"
                      href="/topics?mock=standard"
                      onClick={[Function]}
                      onMouseEnter={[Function]}
                    >
                      Topics
                    </a>
                  </div>
                  <div
                    className="px-3"
                  >
                    /
                  </div>
                </li>
                <li
                  className="flex"
                >
                  <div
                    className="BreadcrumbItem"
                    data-testid="ActiveLink"
                  >
                    <a
                      className="hyperlink"
                      href="/topics/topic-slug?mock=standard"
                      onClick={[Function]}
                      onMouseEnter={[Function]}
                    >
                      topic-name
                    </a>
                  </div>
                  <div
                    className="px-3"
                  >
                    /
                  </div>
                </li>
                <li
                  className="flex"
                >
                  <div
                    className="BreadcrumbItem"
                    data-testid="ActiveLink"
                  >
                    <a
                      className="hyperlink"
                      href="/topics/topic-slug/resources/add?mock=standard"
                      onClick={[Function]}
                      onMouseEnter={[Function]}
                    >
                      Add
                    </a>
                  </div>
                  <div
                    className="px-3"
                  >
                    /
                  </div>
                </li>
                <li
                  className="flex"
                >
                  <div
                    className="BreadcrumbItem"
                    data-testid="ActiveLink"
                  >
                    <a
                      className="hyperlink"
                      href="/topics/topic-slug/resources/new?mock=standard"
                      onClick={[Function]}
                      onMouseEnter={[Function]}
                    >
                      New
                    </a>
                  </div>
                </li>
              </ol>
            </nav>
            <div>
              <div
                className="text-4xl"
              >
                New Topic Resource
              </div>
              <div
                className="mt-10"
              >
                <div
                  className="ResourceDetailsForm"
                  data-testid="ResourceDetailsForm"
                >
                  <form
                    onSubmit={[Function]}
                  >
                    <div>
                      <div
                        className="TextInput"
                        data-testid="TextInput"
                      >
                        <div
                          className="flex flex-col"
                        >
                          <label
                            htmlFor="name"
                          >
                            Name
                          </label>
                          <input
                            autoComplete="off"
                            className="mt-2 form-control flex-grow"
                            id="name"
                            name="name"
                            onBlur={[Function]}
                            onChange={[Function]}
                            type="text"
                          />
                        </div>
                      </div>
                    </div>
                    <div
                      className="mt-6"
                    >
                      <div
                        className="TextInput"
                        data-testid="TextInput"
                      >
                        <div
                          className="flex flex-col"
                        >
                          <label
                            htmlFor="link"
                          >
                            Link
                          </label>
                          <input
                            autoComplete="off"
                            className="mt-2 form-control flex-grow"
                            id="link"
                            name="link"
                            onBlur={[Function]}
                            onChange={[Function]}
                            type="text"
                          />
                        </div>
                      </div>
                    </div>
                    <div
                      className="mt-10 flex flex-col sm:flex-row"
                    >
                      <button
                        className="btn btn-primary w-100"
                        disabled={null}
                        type="submit"
                      >
                        Submit
                      </button>
                      <a
                        className="mt-3 text-center btn btn-cancel-outline w-100 sm:mt-0 sm:ml-3"
                        href="/topics/topic-slug/resources/add?mock=standard"
                        onClick={[Function]}
                        onMouseEnter={[Function]}
                      >
                        Cancel
                      </a>
                    </div>
                  </form>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
      <footer
        className="Footer"
        data-testid="Footer"
      >
        <div
          className="content-container"
        >
          <div
            className="sm:flex text-center sm:text-left"
          >
            <div
              className="sm:order-2 sm:ml-10"
            >
              <h1
                className="text-4xl"
              >
                Our Mission
              </h1>
              <p
                className="py-3"
              >
                is to 
                <strong>
                  simplify
                </strong>
                 your learning journey. Learning can be 
                <strong>
                  difficult
                </strong>
                , especially when starting. There are
                 
                <strong>
                  millions
                </strong>
                 of educational courses, videos, and books
                 
                <strong>
                  scattered
                </strong>
                 across the Internet.
              </p>
              <p
                className="py-3"
              >
                Resource Rankâ€™s goal is to 
                <strong>
                  consolidate
                </strong>
                 all of these resources so that you can 
                <strong>
                  quickly
                </strong>
                 find the
                 
                <strong>
                  best
                </strong>
                 resources for what you are learning.
              </p>
            </div>
            <div
              className="sm:order-1 mt-10 sm:mt-0 flex items-start justify-center sm:justify-start"
            >
              <img
                alt="Resource Rank Logo"
                className="max-w-none"
                height="75"
                src="/img/logo.svg"
                width="75"
              />
            </div>
          </div>
        </div>
      </footer>
    </div>
  </div>
</div>
`;
